llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -21741,45 +21741,63 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
   // The narrower of the two operands. Used as the accumulator
   auto NarrowOp = N->getOperand(1);
   auto MulOp = N->getOperand(2);
```

Comment:
I think the naming of these variables should be reversed; I think the operands coming into the partial.reduce should be named `A` and `B`, whereas the current values for `A` and `B` should be renamed to e.g. `AddOp` and `MulOp`. Because you're trying to match an operation and map that to a dot instruction that extends the operands, where the two operands have a meaning.

For example, after your change in this patch, `MulOp` may no longer be a MUL because it also supports it being an extend.